### PR TITLE
Issue 230 - Adding Stanford Dogs Dataset

### PR DESCRIPTION
**Pull request by iswariyam**
Added files for incorporating Stanford Dogs Dataset [Issue 230](https://github.com/tensorflow/datasets/issues/230#issuecomment-473686949)

https://gist.github.com/iswariyam/f7e3d694b27d2647d6e41a6c482ca423

### CODE REVIEW

**Comment by iswariyam** [comment]
I signed the CLA

**Comment by iswariyam** [comment]
@cyfra @rsepassi @Conchylicultor  Kindly review and let me know if any further changes are required.

File: tensorflow_datasets/image/stanford_dogs.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+            gen_kwargs={
+            "archive": dl_manager.iter_archive(download_path[2]),
+            "file_names": test_list,
+            "annotation_files": xml_file_list,
+            })
+        ]
+        
+    
+    def _generate_examples(self, archive, file_names, annotation_files):
+        """Generate horses or humans images and labels given the directory path
```
Comment (SoumyadeepJana) [inline tensorflow_datasets/image/stanford_dogs.py]: Kindly change the docstring to the dataset name you are adding :P

File: tensorflow_datasets/image/__init__.py [hunk 2]
```diff
@@ -42,10 +42,7 @@
 from tensorflow_datasets.image.open_images import OpenImagesV4
 from tensorflow_datasets.image.quickdraw import QuickdrawBitmap
 from tensorflow_datasets.image.rock_paper_scissors import RockPaperScissors
-<<<<<<< HEAD
 from tensorflow_datasets.image.stanford_dogs import StanfordDogs
```
Comment (SoumyadeepJana) [inline tensorflow_datasets/image/__init__.py]: Arrange the import names alphabetically :)

File: tensorflow_datasets/image/stanford_dogs.py (hunk 1 above)
Comment (iswariyam) [reply tensorflow_datasets/image/stanford_dogs.py]: This has already been done. You are looking at an outdated file.

File: tensorflow_datasets/image/__init__.py (hunk 2 above)
Comment (iswariyam) [reply tensorflow_datasets/image/__init__.py]: Done!

File: tensorflow_datasets/image/stanford_dogs.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                }),
+            }),
+            supervised_keys=("image", "label"),
+            urls=_URL,
+            citation=_CITATION
+        )
+
+    def _split_generators(self, dl_manager):
+
+        download_path = dl_manager.download([_SPLIT_URL, _ANNOTATIONS_URL, _IMAGES_URL])
```
[the remark this answers is not in the document]
Comment (iswariyam) [reply tensorflow_datasets/image/stanford_dogs.py]: Yes, they are needed, because I need the dataset annotation and split files to be extracted out after downloading so that I can parse the contents in them. 
However, the folder containing the images can remain as an archive and can be iterated over using dl_manager.iter_archive() for fetching the examples

File: tensorflow_datasets/image/stanford_dogs.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        bounding box values
+
+        """
+
+        label_names = self.info.features["label"].names
+        bbox_attrib = ["xmin", "xmax", "ymin", "ymax", "width", "height"]
+
+        for fname, fobj in archive:
+            res = _NAME_RE.match(fname)
+            if not res or not fname.split("/")[-1] in file_names:
```
[the remark this answers is not in the document]
Comment (iswariyam) [reply tensorflow_datasets/image/stanford_dogs.py]: This check is for identifying whether an image belongs to the train set or the test set. The names of images in the test set/train set are passed into the function argument file_names. If suppose, an image file name is in neither of the list, then it will be ignored.

File: tensorflow_datasets/image/stanford_dogs.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            file_name = res.group(2)
+            attributes = defaultdict(list)
+            for element in annotation_files[file_name.split(".")[0]].iter():
+                # Extract necessary Bbox attributes from XML file
+                if element.tag.strip() in bbox_attrib:
+                    attributes[element.tag.strip()].append(float(element.text.strip()))
+
+            # BBox attributes in range of 0.0 to 1.0
+            def normalize_bbox(bbox_side, image_side):
+
```
[the remark this answers is not in the document]
Comment (iswariyam) [reply tensorflow_datasets/image/stanford_dogs.py]: Done!

File: tensorflow_datasets/image/stanford_dogs.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        )
+
+    def _split_generators(self, dl_manager):
+
+        download_path = dl_manager.download([_SPLIT_URL, _ANNOTATIONS_URL, _IMAGES_URL])
+        extracted_path = dl_manager.download_and_extract([_SPLIT_URL, _ANNOTATIONS_URL])
+        xml_file_list = defaultdict(str)
+
+        def parse_mat_file(file_name):
+
```
[the remark this answers is not in the document]
Comment (iswariyam) [reply tensorflow_datasets/image/stanford_dogs.py]: Done!

File: tensorflow_datasets/image/stanford_dogs.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+            if "train" in fname: 
+                train_list, train_mat_arr = parse_mat_file(full_file_name)
+                label_names = set([element.split('/')[-2].lower()
+                                  for element in train_mat_arr['file_list']])
+            elif "test" in fname:
+                test_list, _ = parse_mat_file(full_file_name)
+
+        self.info.features["label"].names = tuple(label_names)
+
```
[the remark this answers is not in the document]
Comment (iswariyam) [reply tensorflow_datasets/image/stanford_dogs.py]: I checked it manually and it is 120. But do you want me to put a check here? Also if i do put a check here, what should be my failure case ?

**Comment by iswariyam** [comment]
@cyfra Please let me know if anymore changes are required from my side.
